### PR TITLE
Internal html changes

### PR DESCRIPTION
**Pull request by matthewlipski**
closes https://github.com/TypeCellOS/BlockNote/issues/1064
closes https://github.com/TypeCellOS/BlockNote/issues/1110

### CODE REVIEW

File: packages/core/src/api/exporters/copyExtension.ts [hunk 1]
```diff
@@ -7,9 +7,12 @@ import type { BlockNoteEditor } from "../../editor/BlockNoteEditor";
 import { BlockSchema, InlineContentSchema, StyleSchema } from "../../schema";
 import { initializeESMDependencies } from "../../util/esmDependencies";
 import { createExternalHTMLExporter } from "./html/externalHTMLExporter";
-import { createInternalHTMLSerializer } from "./html/internalHTMLSerializer";
 import { cleanHTMLToMarkdown } from "./markdown/markdownExporter";
 
+// use a dynamic import because we want to access
+// __serializeForClipboard which is not exposed in types
+let pmView: any;
```
Comment (YousefED) [inline packages/core/src/api/exporters/copyExtension.ts:14]: instead of a dynamic import (which also changes the way the module is loaded), I think you can do

```
import * as pmView from "prosemirror-view;
(pmView as any).__serializeForClipboard
```

File: packages/core/src/api/exporters/copyExtension.ts [hunk 2]
```diff
@@ -22,16 +25,40 @@ async function selectedFragmentToHTML<
...
 
-  const internalHTMLSerializer = createInternalHTMLSerializer(
-    view.state.schema,
-    editor
-  );
-  const internalHTML = internalHTMLSerializer.serializeProseMirrorFragment(
-    selectedFragment,
-    {}
-  );
+  const isWithinBlockContent =
```
Comment (YousefED) [inline packages/core/src/api/exporters/copyExtension.ts:48]: is there a test for this fix?

File: packages/core/src/api/exporters/html/__snapshots_fragment_edge_cases__/selectionLeavesBlockChildren.html [hunk 3]
```diff
@@ -1 +0,0 @@
-<p>Nested Paragraph 1</p><p>Nested Paragraph 2</p><p>Nested Paragraph 3</p><p>Paragraph 2</p>
```
Comment (YousefED) [inline packages/core/src/api/exporters/html/__snapshots_fragment_edge_cases__/selectionLeavesBlockChildren.html:1]: why have these been removed?

File: packages/core/src/api/exporters/html/htmlConversion.test.ts [hunk 4]
```diff
@@ -107,144 +116,304 @@ describe("Test HTML conversion", () => {
...
   let editor: BlockNoteEditor;
   const div = document.createElement("div");
+  document.body.append(div);
+
   beforeEach(() => {
+    editor.replaceBlocks(editor.document, initialContent);
+  });
+
+  beforeAll(async () => {
+    pmView = await import("prosemirror-view");
```
Comment (YousefED) [inline packages/core/src/api/exporters/html/htmlConversion.test.ts:255]: same as earlier

File: packages/core/src/api/exporters/html/htmlConversion.test.ts [hunk 5]
```diff
@@ -107,144 +116,304 @@ describe("Test HTML conversion", () => {
...
-      // Selection starts and ends within the first block's children.
+  // Sets the editor selection to the given start and end positions, then
+  // exports the selected content to HTML and compares it to a snapshot.
+  function testSelection(testName: string, startPos: number, endPos: number) {
+    editor.dispatch(
+      editor._tiptapEditor.state.tr.setSelection(
+        TextSelection.create(editor._tiptapEditor.state.doc, startPos, endPos)
+      )
+    );
+    if (
```
Comment (YousefED) [inline packages/core/src/api/exporters/html/htmlConversion.test.ts:280]: does this actually happen? Because you `setSelection` to a `TextSelection`, not a `NodeSelection` which I think doesn't have a node property.

File: packages/core/src/api/exporters/html/htmlConversion.test.ts [hunk 6]
```diff
@@ -107,144 +116,304 @@ describe("Test HTML conversion", () => {
...
 });
 
-// Fragments created from ProseMirror selections don't always conform to the
-// schema. This is because ProseMirror preserves the full ancestry of selected
-// nodes, but not the siblings of ancestor nodes. These tests are to verify that
-// Fragments like this are exported to HTML properly, as they can't be created
-// from Block objects like all the other test cases (Block object conversions
-// always conform to the schema).
-describe("Test ProseMirror fragment edge case conversion", () => {
+// These tests are meant to test the copying of user selections in the editor.
```
Comment (YousefED) [inline packages/core/src/api/exporters/html/htmlConversion.test.ts:119]: extract this stuff to a separate test file? lot's going on now

File: packages/core/src/api/exporters/html/htmlConversion.test.ts [hunk 7]
```diff
@@ -107,144 +116,304 @@ describe("Test HTML conversion", () => {
...
-          TextSelection.create(editor._tiptapEditor.state.doc, 18, 97)
-        )
-      );
+    expect(internalHTML).toMatchFileSnapshot(
+      `./__snapshots_selection_html__/${testName}.html`
+    );
 
-      const copiedFragment =
-        editor._tiptapEditor.state.selection.content().content;
+    const recreatedSlice = pmView.__parseFromClipboard(
```
Comment (YousefED) [inline packages/core/src/api/exporters/html/htmlConversion.test.ts:308]: this basically tests whether `__parseFromClipboard` works as the reverse from `__serializeForClipboard` right? i.e. we're basically testing prosemirror internals.

File: packages/core/src/blocks/TableBlockContent/TableBlockContent.ts [hunk 8]
```diff
@@ -46,14 +46,18 @@ const TableParagraph = Node.create({
 
   parseHTML() {
     return [
-      { tag: "td" },
       {
         tag: "p",
+        priority: 1000,
```
Comment (YousefED) [inline packages/core/src/blocks/TableBlockContent/TableBlockContent.ts:51]: curious about the changes in this file; why were they necessary and are they covered in the tests?

File: packages/core/src/schema/blocks/createSpec.ts [hunk 9]
```diff
@@ -162,15 +162,24 @@ export function createBlockSpec<
...
+      // the editor, so we need to make sure the `blockContent` element is
+      // structured correctly as this is what's used for parsing blocks. We
+      // just render a placeholder div inside as the `blockContent` element
+      // already has all the information needed for proper parsing.
       const div = document.createElement("div");
       div.setAttribute("data-tmp-placeholder", "true");
-      return {
-        dom: div,
-      };
+      return wrapInBlockStructure(
```
Comment (YousefED) [inline packages/core/src/schema/blocks/createSpec.ts:173]: nice :)

**Review by YousefED — COMMENTED** [review]
Sweet, almost there! Couple of small questions

Regarding testing; curious whether the tests we created also work with the original code and if we can see if the original code breaks the tests and the new code fixes it. That way we know for sure our tests + code do what we need :) Let's discuss!

File: packages/core/src/api/parsers/html/parseHTML.test.ts [hunk 10]
```diff
@@ -6,20 +7,18 @@ async function parseHTMLAndCompareSnapshots(
...
   editor.mount(div);
   const blocks = await editor.tryParseHTMLToBlocks(html);
 
-  const snapshotPath = "./__snapshots__/paste/" + snapshotName + ".json";
+  const snapshotPath = "./__snapshots__/" + snapshotName + ".json";
   expect(JSON.stringify(blocks, undefined, 2)).toMatchFileSnapshot(
     snapshotPath
   );
 
+  // TODO: I don't think is really related to parsing. It's paste behaviour
```
Comment (matthewlipski) [inline packages/core/src/api/parsers/html/parseHTML.test.ts:20]: @YousefED thoughts on removing this?

File: packages/core/src/api/clipboard/__snapshots__/childToParent.html [hunk 11]
```diff
@@ -0,0 +1 @@
+<h2 data-text-color="red" data-level="2">Heading 1</h2><p>Nested Paragraph 1</p>
```
Comment (YousefED) [inline packages/core/src/api/clipboard/__snapshots__/childToParent.html:1]: I'm confused; shouldn't these snapshots have the prosemirror slice info?

Comment (YousefED) [reply packages/core/src/api/clipboard/__snapshots__/childToParent.html:1]: Ah wait, you test te externalhtml with snapshots, but the clipboard by calling paste, is that the idea?
